software/main.py

```diff
@@ -0,0 +1,335 @@
+        # stop button
+        if self.controller.is_stopped():
+            raise KeyboardInterrupt
+
+        # verify if robot is remote controlled, no autonomous stuff if robot is remote controlled
+        if self.controller.is_remote_controlled():
+            self.current_state = State.RemoteControl
+        if self.current_state == State.RemoteControl:
+            if not self.controller.is_remote_controlled():
+                self.back_to_search_state()
+            else:
+                return
+
+        # -- AUTONOMOUS STUFF --
+        if self.ball_count == 0:
+            self.no_balls_frames += 1
+        else:
+            self.no_balls_frames = 0
+
+        self.ball_count = len(self.processed_data.balls)
+        if self.prev_ball_count != self.ball_count:
+            # this is for debugging
+            print("ball_count: {}".format(self.ball_count))
+            self.prev_ball_count = self.ball_count
+
+        # first element is always the closest ball
+        if self.ball_count > 0:
+            self.ball = self.processed_data.balls[0]
+        else:
+            self.ball = None
+
+        if self.basket_color == Color.MAGENTA:
+            self.basket = self.processed_data.basket_m
+        elif self.basket_color == Color.BLUE:
+            self.basket = self.processed_data.basket_b
+
+    def searching_state(self):
+        """State for searching for the ball"""
+        if self.ball_count != 0:
+            self.current_state = State.DriveToBall
+
+        elif time() < self.search_end:
+            print("--Searching-- Moving to look for ball")
+            self.robot.move(0, 0, self.search_speed, 0)
+
+        else:
+            print("--Searching-- Entering wait to scan surroundings")
+            self.current_state = State.Wait
+            self.wait_end = time() + self.scan_wait_time
+            self.next_state = State.Searching
+
+    def wait_state(self):
+        """State for waiting"""
+        if self.next_state == State.Searching and self.ball_count != 0:
+            print("--Wait-- Found ball.")
+            self.current_state = State.DriveToBall
+            return
+        if time() >= self.wait_end:
+            if self.next_state is not None:
+                self.current_state = self.next_state
+            else:
+                self.current_state = State.Searching
+            self.next_state = None
+            if self.current_state == State.Searching:
+                self.search_end = time() + self.scan_move_time
+
+    def drive_to_ball_state(self):
+        """State for driving to the ball."""
+        if self.no_balls_frames > self.max_ball_miss:  # lost the ball
+            print(
+                f"--DriveToBall-- Haven't seen ball for {self.max_ball_miss} frames, going back to search.")
+            self.back_to_search_state()
+            return
+        if self.ball_count == 0:  # don't do anything when we cant see a ball
+            self.robot.stop()
```

Comment:
It was there before because we mainly tested with debug = True, where the framerate was super low, so ball locations between 2 frames could be very different, which is not great for following the ball.

Removed now.